### PR TITLE
txn: Reject commit requests with commit_ts < primary_lock.min_commit_ts

### PR DESCRIPTION
**Pull request by MyonKeminta**
###  What have you changed?

This PR rejects commit requests that `commit_ts < primary_lock.min_commit_ts`.

This is part of supporting large transactions. The large transaction's primary lock has a field `min_commit_ts`, which can be updated other transactions, and the other transaction can then read ignoring the lock. If the actual commit_ts is less than the min_commit_ts, the RR isolation level will be broken.

###  What is the type of the changes?

Pick one of the following and delete the others:

- New feature (a change which adds functionality)

###  How is the PR tested?

- Unit test

###  Does this PR affect documentation (docs) or should it be mentioned in the release notes?

No

###  Does this PR affect `tidb-ansible`?

No

###  Refer to a related PR or issue link (optional)

- [x] Required PR in kvproto: https://github.com/pingcap/kvproto/pull/473
- [x] Blocked by: https://github.com/tikv/tikv/pull/5683 (changes in kvproto)

### CODE REVIEW

**Comment by MyonKeminta** [comment]
@tiancaiamao @coocood @youjiali1995 @sticnarf PTAL thanks!

**Comment by coocood** [comment]
LGTM

File: src/storage/mvcc/txn.rs [hunk 1]
```diff
@@ -1109,6 +1125,25 @@ mod tests {
...
+        must_prewrite_put_for_large_txn(&engine, k, v, k, ts(10, 0), 100, 0);
+        must_check_txn_status(&engine, k, ts(10, 0), ts(20, 0), ts(20, 0), 100, 0);
+        // The the min_commit_ts should be ts(20, 1)
+        must_commit_err(&engine, k, ts(10, 0), ts(15, 0));
+        must_commit_err(&engine, k, ts(10, 0), ts(20, 0));
+        must_commit(&engine, k, ts(10, 0), ts(20, 1));
+
+        must_prewrite_put_for_large_txn(&engine, k, v, k, ts(30, 0), 100, 0);
+        must_check_txn_status(&engine, k, ts(30, 0), ts(40, 0), ts(40, 0), 100, 0);
+        must_commit(&engine, k, ts(30, 0), ts(50, 0));
```
Comment (sticnarf) [inline src/storage/mvcc/txn.rs]: What are these lines for?

File: src/storage/mvcc/txn.rs [hunk 2]
```diff
@@ -458,6 +458,22 @@ impl<S: Snapshot> MvccTxn<S> {
                         pessimistic: true,
                     });
                 }
+                // A lock with larger min_commit_ts than current commit_ts can't be committed
+                if commit_ts < lock.min_commit_ts {
+                    error!(
```
Comment (sticnarf) [inline src/storage/mvcc/txn.rs]: Not necessary to log with error

Comment (MyonKeminta) [reply src/storage/mvcc/txn.rs]: Oh, sorry.
Is info sufficient?

File: src/storage/mvcc/txn.rs (hunk 1 above)
Comment (MyonKeminta) [reply src/storage/mvcc/txn.rs]: I'm trying to cover `commit_ts < min_commit_ts`, `commit_ts == min_commit_ts` and  `commit_ts > min_commit_ts`, and this is the last case.

File: src/storage/mvcc/txn.rs (hunk 2 above)
Comment (sticnarf) [reply src/storage/mvcc/txn.rs]: I think info is OK. It's normal that this case happens.

File: src/storage/mvcc/txn.rs (hunk 1 above)
Comment (sticnarf) [reply src/storage/mvcc/txn.rs]: Got it.

**Comment by zhouqiang-cl** [comment]
there are many coredump @MyonKeminta

**Comment by MyonKeminta** [comment]
@zhouqiang-cl The kvproto's update includes https://github.com/pingcap/kvproto/pull/467 , which changes fields about checksum, and it is used by https://github.com/tikv/tikv/pull/5683 . I have no choice other than temporarily commenting the related code out to make this PR compile, and wait for merging #5683 before merging this RP.

**Comment by zhouqiang-cl** [comment]
got

**Comment by MyonKeminta** [comment]
@youjiali1995 @sticnarf PTAL

File: src/storage/mvcc/txn.rs [hunk 3]
```diff
@@ -458,6 +458,22 @@ impl<S: Snapshot> MvccTxn<S> {
                         pessimistic: true,
                     });
                 }
+                // A lock with larger min_commit_ts than current commit_ts can't be committed
+                if commit_ts < lock.min_commit_ts {
```
Comment (sticnarf) [inline src/storage/mvcc/txn.rs]: What if the cluster is upgraded from an old version and its `min_commit_ts` is 0?

**Comment by MyonKeminta** [comment]
Oh, sorry. I forgot to remove the DNM tag ..

**Pull request merged**